src/reflect/scala/reflect/io/PlainFile.scala

```diff
@@ -115,7 +115,29 @@ private[scala] class PlainNioFile(nioPath: java.nio.file.Path) extends AbstractF
 
   override lazy val canonicalPath = super.canonicalPath
 
-  override def underlyingSource  = Some(this)
+  override def underlyingSource  = {
+    val fileSystem = nioPath.getFileSystem
+    fileSystem.provider().getScheme match {
+      case "jar" =>
+        val fileStores = fileSystem.getFileStores.iterator()
+        if (fileStores.hasNext) {
+          val jarPath = fileStores.next().name
+          try {
+            Some(new PlainNioFile(Paths.get(jarPath.stripSuffix(fileSystem.getSeparator))))
+          } catch {
+            case _: InvalidPathException =>
+              None
+          }
+        } else None
+      case "jrt" =>
+        if (nioPath.getNameCount > 2 && nioPath.startsWith("/modules")) {
+          // TODO limit this to OpenJDK based JVMs?
+          val moduleName = nioPath.getName(1)
+          Some(new PlainNioFile(Paths.get(System.getProperty("java.home"), "jmods", moduleName.toString + ".jmod")))
```

Comment:
Does this hold true for custom modules?

Reply:
Previewing of this patch together with my review branch of https://github.com/scala/scala/pull/7218 shows that things should work. We end up in the `"jar"` case above for custom modules, the `"jrt"` case is just for the modules bundled with the JDK. 

```
$ ./build/quick/bin/scala -nobootcp -Yjpms -modulepath $(coursier fetch -q -p org.apache.commons:commons-rng-simple:1.2) -addmodules org.apache.commons.rng.simple

scala> :power

scala> val cls = rootMirror.getClassByName(TermName("org.apache.commons.rng.simple.RandomSource"))
cls: $r.intp.global.ClassSymbol = Java enum RandomSource

scala> cls.initialize.associatedFile
res7: scala.reflect.io.AbstractFile = /org/apache/commons/rng/simple/RandomSource.class

scala> .underlyingSource
res8: Option[scala.reflect.io.AbstractFile] = Some(/Users/jz/.coursier/cache/v1/https/repo1.maven.org/maven2/org/apache/commons/commons-rng-simple/1.2/commons-rng-simple-1.2.jar)

scala> val cls = rootMirror.getClassByName(TermName("java.lang.String"))
cls: $r.intp.global.ClassSymbol = class String

scala> cls.initialize.associatedFile
res9: scala.reflect.io.AbstractFile = /modules/java.base/java/lang/String.class

scala> .underlyingSource
res10: Option[scala.reflect.io.AbstractFile] = Some(/Users/jz/.jabba/jdk/openjdk@1.11.0/Contents/Home/jmods/java.base.jmod)
```

Reply:
Nice.